chore: add runtime component deprecation warnings

jcfranco opened the pull request:
**Related Issue:** #9295

## Summary

This adds deprecation warning messages to components marked for deprecation at v3 and v4.

driskull commented on packages/calcite-components/src/components/pick-list-group/pick-list-group.tsx:
```diff
@@ -6,8 +6,15 @@ import {
+import { logger } from "../../utils/logger";
 import { CSS, SLOTS } from "./resources";
 
+logger.deprecated("component", {
```
this is super cool!

driskull commented on packages/calcite-components/src/utils/logger.ts:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  const badgeTemplate = "%ccalcite";
+  const badgeStyle = "background: #007AC2; color: #fff; border-radius: 4px;padding: 2px 4px;";
```
Should design approve the styling :)

Wondering if we could use CSS vars here instead of styles. I guess it doesn't really matter since its just a log badge but still.

jcfranco replied:
I did test this, but no dice. Custom CSS props don't appear to be available for console styling. 😞

driskull commented on packages/calcite-components/src/utils/logger.ts:
```diff
@@ -0,0 +1,84 @@
+    listFormatter = new Intl.ListFormat("en", { style: "long", type: "disjunction" });
+  }
+
+  const message = `[${name}] ${context} is deprecated and will be removed in ${removalVersion === "future" ? `a future version` : `v${removalVersion}`}.${suggested ? ` Use ${multiSuggestions ? listFormatter.format(suggested.map((suggestion) => `"${suggestion}"`)) : `"${suggested}"`} instead.` : ""}`;
```
We don't need to translate this right?

jcfranco replied:
Correct. Dev messaging is not translated.

driskull commented:
1. I do like the badge
2. I don't think we need the calcite prefix as it is redundant with the badge.
3. I think its probably fine in the config for now. What do you think?

driskull reviewed (approved):
This will be awesome! 💯

DitwanP commented:
This is great! 💪 👏

I'm with Matt on points 1, 2, and 3 from his first comment above.

geospatialem reviewed (commented):
**_Awesome!!_** 💪🏻 

As far as the version, should we also mention a timeframe for folks to plan, or keep it vague with the version? Something like:

> `calcite` [tip-group] component is deprecated and will be removed in v4 (2025 Q4). Use "carousel" or "carousel item" instead.

> 1. Badge 👍👎? 

Badge it up! 🦡📛 
 
> 2. Assuming 👍 on the badge, should we keep calcite in component names used in messaging? The badge already provides context

For consistency with our documentation, its okay to omit since its mentioned already in the badge. If we remove the badge, we should leave it in though. 

The only hesitation would be if omitting it creates confusion for developers to use the recommended component(s).

It looks like the messaging also adds `"` to the recommended components, should the symbol be around each of the recommendations, like:

> Use "carousel" or "carousel-item" instead.

> 3. Do we want to keep logLevel in the config or should we do something similar to [Maps SDK's config](https://developers.arcgis.com/javascript/latest/api-reference/esri-config.html#log) for future log-related features? I don't see a use case for interceptors at the moment.

Flexible on whichever approach is best as long as it is clear to developers working with both Maps SDK and Calcite.

benelan reviewed (approved):
I like this idea as well! One suggestion that hasn't been mentioned, maybe add a "none" or "disabled" option to `logLevels`? That way teams can completely turn it off in prod and set it to something verbose in dev.

jcfranco commented:
> As far as the version, should we also mention a timeframe for folks to plan, or keep it vague with the version? Something like:

I like the idea of adding the timeframe, but this would be tricky to maintain and enforce via TypeScript. 

> The only hesitation would be if omitting it creates confusion for developers to use the recommended component(s).

I hear ya. We should be good on this front since we do have the badge giving context and often omit `calcite` in the doc when referring to components. This also makes the message less noisy.

> It looks like the messaging also adds " to the recommended components, should the symbol be around each of the recommendations, like:

My bad. 😅 I used an outdated screenshot. Each component should be quoted in the latest version (covered by a test too).

jcfranco commented:
Thanks for the feedback, y'all! 🚀 Made the following changes:

* dropped namespace from component messaging to simplify (this is enforced by TS)
* added an `off` level to mute all logs

eriklharper commented:
+1 for badge, +1 for keeping `calcite` in the name of the component message.  I think that's clearer for folks outside the core team since that's how they interact with the components.  Keeping log level in the config sounds like a good first step, we can always add more to that down the road.
